pkg/build/webhook/gitlab/gitlab_test.go

```diff
@@ -0,0 +1,363 @@
+	}
+}
+
+func TestJsonGitLabPushEvent(t *testing.T) {
+	req := postFile("X-Gitlab-Event", "Push Hook", "pushevent.json", "http://some.url", http.StatusOK, t)
+	plugin := New()
+	_, _, _, proceed, err := plugin.Extract(buildConfig, "secret100", "", req)
+
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	if !proceed {
+		t.Error("Expected 'proceed' return value to be 'true'")
+	}
+}
+
+func TestJsonGitLabPushEventWithCharset(t *testing.T) {
+	req := postFileWithCharset("X-Gitlab-Event", "Push Hook", "pushevent.json", "http://some.url", "application/json; charset=utf-8", http.StatusOK, t)
+	plugin := New()
+	_, _, _, proceed, err := plugin.Extract(buildConfig, "secret100", "", req)
+
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	if !proceed {
+		t.Error("Expected 'proceed' return value to be 'true'")
+	}
+}
+
+func postFile(eventHeader, eventName, filename, url string, expStatusCode int, t *testing.T) *http.Request {
+	return postFileWithCharset(eventHeader, eventName, filename, url, "application/json", expStatusCode, t)
+}
+
+func postFileWithCharset(eventHeader, eventName, filename, url, charset string, expStatusCode int, t *testing.T) *http.Request {
+	data, err := ioutil.ReadFile("testdata/" + filename)
+	if err != nil {
+		t.Errorf("Failed to open %s: %v", filename, err)
+	}
+
+	return postWithCharset(eventHeader, eventName, data, url, charset, expStatusCode, t)
+}
+
+func post(eventHeader, eventName string, data []byte, url string, expStatusCode int, t *testing.T) *http.Request {
+	return postWithCharset(eventHeader, eventName, data, url, "application/json", expStatusCode, t)
+}
+
+func postWithCharset(eventHeader, eventName string, data []byte, url, charset string, expStatusCode int, t *testing.T) *http.Request {
+	req, err := http.NewRequest("POST", url, bytes.NewReader(data))
+	if err != nil {
+		t.Errorf("Error creating POST request: %v", err)
+	}
+
+	req.Header.Add("Content-Type", charset)
+	req.Header.Add(eventHeader, eventName)
+
+	return req
+}
+
+type testContext struct {
+	plugin   WebHook
+	buildCfg *api.BuildConfig
+	req      *http.Request
+	path     string
+}
+
+func setup(t *testing.T, filename, eventType, ref string) *testContext {
+	context := testContext{
+		plugin: WebHook{},
+		buildCfg: &api.BuildConfig{
+			Spec: api.BuildConfigSpec{
+				Triggers: []api.BuildTriggerPolicy{
+					{
+						Type: api.GitLabWebHookBuildTriggerType,
+						GitLabWebHook: &api.WebHookTrigger{
+							Secret: "secret101",
+						},
+					},
+					{
+						Type: api.GitLabWebHookBuildTriggerType,
+						GitLabWebHook: &api.WebHookTrigger{
+							Secret: "secret100",
+						},
+					},
+					{
+						Type: api.GitLabWebHookBuildTriggerType,
+						GitLabWebHook: &api.WebHookTrigger{
+							Secret: "secret102",
+						},
+					},
+				},
+				CommonSpec: api.CommonSpec{
+					Source: api.BuildSource{
+						Git: &api.GitBuildSource{
+							URI: "git://github.com/my/repo.git",
```

Comment:
I think this can be any URI? I think the intended use is a custom gitlab server.